File: segtree.cpp
/** @file segtree.cpp
 *  @author Danny Sleator, translated by Abi Kim (abigalek)
 *  @brief Segment tree implementation in C++. 
 *  Implements the segtree described in the lectures notes.
 */ 


#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Segtree {
    public:
        // Constructor for segtrees.
        Segtree(int size) {
            n = size;
            N = SuperCeiling(size);
            A.resize(2 * N, identity);
        }

        // Implements Assign in the lecture notes.
        // Assigns value x to index i in the array A, and also recomputes
        // every other value in the path up to the root from index i by 
        // adding together the values of the two children below it.
        void Assign(int i, int x) {
            i = i+N;
            A[i] = x;
            for (i = Parent(i); i>0; i = Parent(i)) {
                A[i] = glue (A[LeftChild(i)], A[RightChild(i)]);
            }
        }

        // Finds sum_{i <= k <= j} A[k]
        // More information in function f's documentation.
        int RangeSum(int i, int j) {
            return f (1, 0, (N-1), i, j);
        }

        // Prints the contents of the array A
        void printA() {
            for (int i=0; i < 2*N; i++) {
                cout << "A[" << i << "] = " << A[i] << endl;
            }
        }

    private:
        int n;
        int N;
        vector<int> A;
        int identity = 0;

        // Returns smallest power of 2 greater than equal to n
        int SuperCeiling(int n) {
            int p;
            for (p=1; p < n; p = p << 1);
            return p;
        }

        // Tree index functions
        int Parent(int i) {return i/2;}
        int LeftChild(int i) {return 2*i;}
        int RightChild(int i) {return 2*i+1;}

        // Function that glues associative operator on elements
        int glue(int a, int b) {
            return a+b;
        }

        int f (int v, int l, int r, int i, int j) {
        /*  We’re currently at A[v]. 1 <= v < 2*N.
            The range [l,r] is that of the current block, wrt user variables [0,n-1].
            The range [i,j] is the range of the query, wrt user variables [0,n-1].
            The size of the range [l,r] = r-l+1 is a power of 2.
            The range [l,r] contains the range [i,j].
            This function returns the answer to the query.
        */
            int t1, t2, m;
            if (l==i && r==j) {
                return A[v];
            } else {
                m = (l+r)/2; // split [l,r] into [l,m] [m+1,r]
                t1 = (i <= m)? f (LeftChild(v), l, m, i, (min(m,j))): identity;
                t2 = (j > m)? f (RightChild(v), (m+1), r, (max(i,(m+1))), j): identity;
                return glue (t1, t2);
            }
        }
};

int main() {
    Segtree s(7);
    s.Assign(3,7);
    s.Assign(4,1);

    s.printA();
    
    printf(" RangeSum(2,7) = %d\n", s.RangeSum(2,7));
    printf(" RangeSum(0,3) = %d\n", s.RangeSum(0,3));
    printf(" RangeSum(4,5) = %d\n", s.RangeSum(4,5));
    printf(" RangeSum(5,5) = %d\n", s.RangeSum(5,5));
    return 0;
}

File: segtree.rs
// @file segtree.rs
// @author Danny Sleator, translated by Abi Kim (abigalek)
// @brief Segment tree implementation in Rust.
// Implements the segtree described in the lectures notes.

// Module that you use to access the segment tree data structure.
mod segtree_mod {
    pub struct SegmentTree {
        n : i32,
        arr : Vec<i32>,
        identity : i32
    }

    fn super_ceiling(nn : i32) -> i32 {
        let mut p = 1;
        while p < nn {
            p = 2 * p;
        }
        return p;
    }

    // Call new from this module to construct a segment tree.
    pub fn new(size : i32) -> SegmentTree {
        return SegmentTree {
            n : super_ceiling(size),
            arr : vec![0; (2 * super_ceiling(size)) as usize],
            identity : 0
        };
    }

    impl SegmentTree {
        // Implements Assign in the lecture notes.
        // Assigns value x to index i in the array A, and also recomputes
        // every other value in the path up to the root from index i by 
        // adding together the values of the two children below it.
        pub fn assign(&mut self, i : i32, x : i32) -> () {
            let mut i_prime = i + self.n;
            self.arr[i_prime as usize] = x;
            i_prime = SegmentTree::parent(i_prime);
            while i_prime > 0 {
                self.arr[i_prime as usize] = SegmentTree::glue (self.arr[SegmentTree::left_child(i_prime) as usize], self.arr[SegmentTree::right_child(i_prime) as usize]);
                i_prime = SegmentTree::parent(i_prime);
            }
        }

        // Finds sum_{i <= k <= j} A[k]
        // More information in function f's documentation.
        pub fn range_sum(&self, i : i32, j : i32) -> i32 {
            return self.f (1, 0, self.n - 1, i, j);
        }

        // Prints the contents of the array arr
        pub fn print_a(&self) -> () {
            for i in 0..self.arr.len() {
                println!("A[{}] = {}", i, self.arr[i]);
            }
        }

        fn parent(i : i32) -> i32 {
            return i/2;
        }

        fn left_child(i : i32) -> i32 {
            return 2 * i;
        }

        fn right_child(i : i32) -> i32 {
            return (2 * i) + 1;
        }

        fn glue (a : i32, b : i32) -> i32 {
            return a + b;
        }

        fn f (&self, v : i32, l : i32, r : i32, i : i32, j : i32) -> i32 {
        // We’re currently at A[v]. 1 <= v < 2*N.
        // The range [l,r] is that of the current block, wrt user variables [0,n-1].
        // The range [i,j] is the range of the query, wrt user variables [0,n-1].
        // The size of the range [l,r] = r-l+1 is a power of 2.
        // The range [l,r] contains the range [i,j].
        // This function returns the answer to the query.

            if l == i && r == j {
                return self.arr[v as usize];
            } else {
                let m = (l+r)/2; // split [l,r] into [l,m] [m+1,r]
                let t1 = if i <= m { self.f(SegmentTree::left_child(v), l, m, i, std::cmp::min(m,j)) } else { self.identity };
                let t2 = if j > m { self.f(SegmentTree::right_child(v), m+1, r, std::cmp::max(i,m+1), j) } else { self.identity };
                return SegmentTree::glue (t1, t2);
            }
        }
    }
}

fn main() {
    let mut s = segtree_mod::new(7);
    s.assign(3,7);
    s.assign(4,1);
    s.print_a();

    println!("range_sum(2, 7) = {}", s.range_sum(2, 7));
    println!("range_sum(0, 3) = {}", s.range_sum(0, 3));
    println!("range_sum(4, 5) = {}", s.range_sum(4, 5));
    println!("range_sum(5, 5) = {}", s.range_sum(5, 5));
}